Razor Pages Support for Admin

ns8482e opened the pull request:
Fixes #3739 by Adding `[Admin]` attribute to `PageModel`

**Pages/Admin/AdminHello.cshtml**

```
@page
@model DemoModel
@using Microsoft.AspNetCore.Mvc.RazorPages
@using OrchardCore.Admin;

@functions
{
    [Admin]
    public class DemoModel : PageModel
    {
        public string Title { get; set; }
        public string Message { get; private set; }

        public void OnGet()
        {
            Title = "This is a RazorPage coming from OrchardCore.Demo!";
            Message = $"Date and time on the server: { DateTime.Now.ToString() }";
        }
    }
}

@await Html.PartialAsync("Header", Model)
<h2>Hello from @ViewContext.RouteData.Values["page"]</h2>
<p>
    @T[Model.Message]
</p>
<p>
    Date using the DateTimeShape: @await DisplayAsync(await New.DateTime(Utc: null, Format: T["MMMM dd, yyyy"].Value))
</p>

```

![image](https://user-images.githubusercontent.com/23270244/72192202-65a5da80-33ca-11ea-9914-cd1a9dbea6c8.png)

jtkech commented:
And i assume we don't need an ordered filter because resource filters run before result ones.

So LGTM

deanmarcussen commented:
I think we need to update the `AdminFilter` to do the same for the permissions check

deanmarcussen commented:
you could probably put that demo code above in the `OrchardCore.Demo` project as well if you wanted @ns8482e 

more examples are good ;)

ns8482e commented:
@deanmarcussen, I tried, but I could not make it work with AdminUrlPrefix and I thought not to include incomplete example 

I created `Pages/Admin/AdminHello.cshtml`  and I tried to map URL similar to AdminController so that page can be accessible using `'/Admin/Demo/AdminHello'`  but URL mapped as `'Admin/Demo/Admin/AdminHello'`

```
 // Add a custom folder route
options.Conventions.AddAreaFolderRoute("OrchardCore.Demo", "/Admin", _adminOptions.AdminUrlPrefix + "/Demo");

```
I could not remove folder name (Admin) from the url

deanmarcussen commented:
cool @ns8482e it's the `AdminFilter` that's important here

ns8482e commented:
> cool @ns8482e it's the `AdminFilter` that's important here

I agree, need to create AdminPageFilter and AdminFolderFilter for Razor Pages similar to Action Filters like  AdminFilter and AdminZoneFilter.

ns8482e commented:
Do we really need `AdminZoneFilter` ? - It apply `AdminAttribute `if controller is `AdminController`. However `AdminFilter` already take care of this with `IsNameAdmin()` condition.

@jtkech @deanmarcussen

jtkech commented:
@ns8482e i did some tests but i didn't fork your branch so i committed a draft PR.

So @ns8482e @deanmarcussen what i'm saying here can be seen here #5259.

- For testing the execution order i kept `AdminAttribute` an `ActionFilterAttribute` with its `Order = -1000` and then just added `IAsyncPageFilter`. I didn't add `IOrderedFilter` because the action filter attribute is already one, then i checked that the `Order` is taken into account on a page action. Notice that when a filter is added through the `MvcOptions` this is at this time that the order is set with the value you provide or 0 by default, but for an attribute filter this the order value defined in your class.

- So yes, this is `AdminAttribute` and `AdminZoneFilter` that apply the admin attribute, one effectively by the attribute the other by the controller name, and they have both an order of -1000, one  through the class definition, the other when added to the mvc filters in the `OC.Admin` startup.

    @deanmarcussen in `AdminZoneFilter` i changed the way we check the controller name, as we did in #5225, by checking if it is an `ControllerActionDescriptor`, maybe not good to do this limitation, but for our mvc admin actions, if it is not a page action, i don't see what else it could be.

    Then i just check if `descriptor.ControllerName == "Admin" where mvc has removed for us the `Controller` part, but what we would need to do for `AdminAbcdController`, hmm i think that's okay but not sure with `adminController` starting with a lower case letter, maybe at least redo a case insensitive check, as you want.

    Then i made `AdminZoneFilter` also an `IPageFilter` that check the action name like this

      public void OnPageHandlerExecuting(PageHandlerExecutingContext context)
      {
          if (Path.GetFileNameWithoutExtension(context.ActionDescriptor.RelativePath).StartsWith("Admin", StringComparison.OrdinalIgnoreCase))
          {
              AdminAttribute.Apply(context.HttpContext);
          }
      }

- I made `AdminFilter` also an `IAsyncPageFilter` and i removed the `IsNameAdmin(context)` check because already done by the `AdminZoneFilter`, anyway the check was wrong beause it was checking **if the name of the controller type** (with the `Controller` part) is eaqual to `Admin`.

- I updated `AdminMenuFilter` as in this PR.

- For testing i created 2 demo admin razor pages, one with the admin attribute and one whose name starts with "Admin".

- Finally, about `PageConventionCollection.AddAreaFolderRoute()` there was a little issue, the folder path was not used, see in my draft PR the final code. Then i was able to define the following routes

      options.Conventions.AddAreaFolderRoute("OrchardCore.Demo", "/Admin", _adminOptions.AdminUrlPrefix + "/Demo");

      => /admin/demo/adminhello
         /admin/demo/helloadmin

    Or with individual routes

      options.Conventions.AddAreaPageRoute("OrchardCore.Demo", "/Admin/AdminHello", _adminOptions.AdminUrlPrefix + "/Demo/AdminHello");
      options.Conventions.AddAreaPageRoute("OrchardCore.Demo", "/Admin/HelloAdmin", _adminOptions.AdminUrlPrefix + "/Demo/HelloAdmin");

      => /admin/demo/adminhello
         /admin/demo/helloadmin

      options.Conventions.AddAreaPageRoute("OrchardCore.Demo", "/Admin/AdminHello", _adminOptions.AdminUrlPrefix + "/AdminHello");
      options.Conventions.AddAreaPageRoute("OrchardCore.Demo", "/Admin/HelloAdmin", _adminOptions.AdminUrlPrefix + "/HelloAdmin");

      => /admin/adminhello
         /admin/helloadmin

ns8482e commented:
@jtkech  Thank you for detailed explanation, now i understand why `AdminZoneFilter`  is important.

I added `AdminPageFilter` for RazorPages to protect admin RazorPages. 

I kept `AdminAttribute `as `IAsyncResourceFilter` and I also changed `AdminZoneFilter` to be `IAsyncResourceFilter` as these filters apply `AdminAttribute`  and I think that should be done immediately after authentication filters - as  Admin Theme Selector is also looking for `AdminAttribute`.

I also added a condition on `AdminZoneFilter` to apply `AdminAttribute` if Razor Page is from `Pages/Admin` folder

Two Sample Pages 

`/Pages/Admin/Hello.cshtml` - Protected without having [Admin] because its in Admin Folder.
`/Pages/OutsideAdmin.cshtml` - Protected using [Admin]

jtkech commented:
Okay no problem i will look at it tomorrow.

- Please look also at my draft PR #5259 to see what i did so that you could check what you missed or what i missed ;) I also fixed some little issues as our `AddAreaFolderRoute()` that was not using the `folderPath` and how we check the `ControllerName` ... so you can integate them in your PR.

- Hmm, in my draft PR #5259 i'm not using the convention that a page action is an admin one because under the admin folder. Maybe you're right but i thought that a razor page is a kind of controller (in fact a file path based controller) that have page handlers being actions. So i rather checked if the file name e.g `AdminHello` starts with `Admin`. But on another side when using the `asp-page` (see below) we use what we call the page name that is e.g `/Admin/AdminHello` regardless all the registered routes. Will see tomorrow.

    If we use your convention in my draft PR, in place of doing this in `AdminZoneFilter`.

      if (Path.GetFileName(context.ActionDescriptor.DisplayName).StartsWith("Admin"))

    We would have to do something like this

      if (context.ActionDescriptor.DisplayName.StartsWith("/Admin/"))

    This if we don't want to check an admin folder at any level e.g `/Some/Path/Admin/Hello`.

---

So just did some tests about link generation using the anchor helper with e.g `asp-page="/admin/helloadmin"`  where `/admin/helloadmin` is the page name regardless all the routes we have registered for it. So we have the default ones registered by `MapRazorPages()`, those by the page directive e.g `@page "/adminhello"`, the ones with the mvc `AddAreaPageRoute()` and with our `AddAreaFolderRoute()`.

- `AddAreaPageRoute()` wins, then `page "/..."`, then `AddAreaFolderRoute()` and the last processed are the routes registered by `MapRazorPages()`.

- For 2 `AddAreaPageRoute()` the last wins, idem for 2 `AddAreaFolderRoute()` because we set `SuppressLinkGeneration = true`.

**So all seems to be ok, but maybe we don't want that `MapRazorPages()` registers for admin page actions some default routes that doesn't start with the admin prefix, as we are doing in #5225 for mvc controller actions**.

ns8482e commented:
>  I also fixed some little issues as our `AddAreaFolderRoute()` that was not using the `folderPath`

I guess this would break URLs for some people for frontend

If user has following files

`/Pages/Account/List.cshtml`
`/Pages/Contact/Me.cshtml `

configured as
```
options.Conventions.AddAreaFolderRoute("OrchardCore.Demo", "/Account", "DemoAccount");
options.Conventions.AddAreaFolderRoute("OrchardCore.Demo", "/Contact", "DemoContact");
```

Before `AddAreaFolderRoute()` change

`/DemoAccount/Account/List`
`/DemoContact/Contact/Me`

After `AddAreaFolderRoute()` change

`/DemoAccount/List`
`/DemoContact/Me`

jtkech commented on src/OrchardCore.Modules/OrchardCore.Admin/AdminZoneFilter.cs:
```diff
@@ -1,29 +1,34 @@
+            else if(context.ActionDescriptor is PageActionDescriptor)
+            {
+                if( ((PageActionDescriptor)context.ActionDescriptor).ViewEnginePath.StartsWith("/Admin", StringComparison.OrdinalIgnoreCase) )
+                {
```
So here to see if it is an admin page action, you check that it is under the 1st level admin folder, this in place of checking the razor page file name (without extensions) e.g `AdminHello`, i'm fine with this, @deanmarcussen ?

Then i think that here also we can be case insensitive, @deanmarcussen ?

Here in my PR i ended up to use `descriptor.DisplayText` that was always with the right letter cases e.g `/Admin/AdminHello`, i don't remember with `ViewEnginePath` but if it's the case that's okay.

So finally you could just use

    if (pageDescriptor.ViewEnginePath.StartsWith("/Admin"))

deanmarcussen replied:
yes I agree we can go case insensitive here, and also that we go with the admin folder name convention, rather than checking the razor page file name. seems to make sense to me.

ns8482e replied:
I guess having Admin folder ( `/Admin/` ) at any level is good enough

jtkech replied:
@deanmarcussen @ns8482e okay cool

@ns8482e then just use `if (pageDescriptor.ViewEnginePath.Contains("/Admin/"))`

ns8482e replied:
@jtkech  @deanmarcussen Here, For controller, I wasn't sure if you want me to change following 

`action.ControllerName.StartsWith("Admin", StringComparison.OrdinalIgnoreCase)`

To, 

`action.ControllerName == "Admin"`

jtkech replied:
`action.ControllerName == "Admin"`

ns8482e commented on src/OrchardCore.Modules/OrchardCore.Admin/AdminZoneFilter.cs:
```diff
@@ -1,29 +1,34 @@
         {
-            var controller = context.Controller as Controller;
-            if (controller != null)
+            if(context.ActionDescriptor is ControllerActionDescriptor)
```
Will do

ns8482e commented on src/OrchardCore.Modules/OrchardCore.Admin/AdminZoneFilter.cs:
```diff
@@ -1,29 +1,34 @@
                     AdminAttribute.Apply(context.HttpContext);
                 }
             }
+            else if(context.ActionDescriptor is PageActionDescriptor)
```
Will do

ns8482e commented on src/OrchardCore.Modules/OrchardCore.Demo/Startup.cs:
```diff
@@ -102,6 +102,12 @@ public override void ConfigureServices(IServiceCollection services)
+
+                // Add a custom page route
+                options.Conventions.AddAreaPageRoute("OrchardCore.Demo", "/OutsideAdmin", _adminOptions.AdminUrlPrefix + "/Demo/OutsideAdmin");
+
```
For admin, I am thinking to create `IPageRouteModelConvention` that would remove `/Admin/` and prefix all admin route with `AdminOption` - So need to specify individual route for Admin.

jtkech replied:
Okay good idea, but, as in #5225 for controller actions, we will have to check that for an admin page action there is no more **by default** any valid route without the admin prefix, and that we still allow people to provide routes (with or without the prefix) through modules that will have an higher precedence for link generaration.

For controller actions, the default routes are registered in the `OC.Mvc startup Configure()` that runs last so that people can register routes with an higher precedence. In #5225 this is there that admin actions get other default routes with the admin prefix.

ns8482e commented on src/OrchardCore.Modules/OrchardCore.Demo/Startup.cs:
```diff
@@ -102,6 +102,12 @@ public override void ConfigureServices(IServiceCollection services)
 
+                 // Add a custom folder route
+                options.Conventions.AddAreaFolderRoute("OrchardCore.Demo", "/Admin", _adminOptions.AdminUrlPrefix + "/Demo");
+
```
With `IPageRouteModelConvention`  that I am currently working, it will convert 
`OrchardCore.Demo/Admin/Hello`
`OrchardCore.Demo/Account/Admin/Create`

To,
`Admin/OrchardCore.Demo/Hello`
`Admin/OrchardCore.Demo/Account/Create`

ns8482e replied:
>i think we need to fix `AddAreaFolderRoute()`
Sure, will change! . However I have one qq, why it's adding new `SelectorModel` and also keeping the original and not replacing?

jtkech replied:
Because `AddAreaFolderRoute()` adds a route ;)

We would need a `ReplaceAreaFolderRoute()` but i don't think so, we would need to have the same behavior as for controller actions where all registered routes are still valid but the first registered routes have an higher precedence for link generation. That's why the `OC.Mvc startup Configure()` run last to register all defaults routes so that we can register routes through modules that will have an higher precedence.

So e.g, if you use an `IPageRouteModelConvention` for admin page actions, we would have to check that modules can still register routes with an higher precedence for link generation.

ns8482e commented:
Also, added `Pages/Foo/List.cshtml` and `Pages/Foo/Admin/Edit.cshtml` Pages in Demo. 

Anonymous user can create and view`Foo` using `/Demo/Foo/List`.
However for Edit/Delete - requires Admin Access
![foo](https://user-images.githubusercontent.com/23270244/72373408-d7816b00-36cd-11ea-97c1-51fdca079a16.gif)
.

jtkech commented on src/OrchardCore/OrchardCore.Mvc.Core/RazorPages/PageConventionCollectionExtensions.cs:
```diff
@@ -17,13 +17,14 @@ public static class PageConventionCollectionExtensions
-                            .Template.Substring(areaName.Length).TrimStart('/')).TrimEnd('/');
-
+                        var templatePart = selector.AttributeRouteModel.Template.Split(path);                        
+                        var template = templatePart[0].TrimEnd('/') + '/' + folderRoute.Trim('/') + '/' + templatePart[1].TrimStart('/');                        
```
So here it seems that the template will still start with the `area + folderPath` segments, then `folderRoute` segments, then the remaining segments of the origonal template, i think this is wrong, it should be `folderRoute + remaining segments` without `area + folderPath` as done and tried with my code, so maybe better to just keep my code ;)

Note: Normally we avoid overuse of string interpollations in hot paths, maybe also to allocate an array with `Split()`, but here in a convention that runs once that's okay.

Hmm maybe good to use `PathString` here because of its + operator that take care of double slashes, but `PathString` has its traps. Above you used `AttributeRouteModel.CombineTemplates`, is it an mvc extension? Maybe useful here. But that's okay to keep it simple with string operations.

ns8482e replied:
I guess CombineTemplates takes care of '/'

> : Normally we avoid overuse of string interpollations in hot paths

Yes, I prefer `$"{X}/{Y}"`, however here I have tried to stay aligned with your code. 

> think this is wrong, it should be `folderRoute + remaining segments` without `area + folderPath` as done and tried with my code, so maybe better to just keep my code ;)

Your code worked until I added, 
`public class AdminFolderModelConvention : IPageRouteModelConvention`

Template no longer starting with `areaName + folderPath ` now it starts with `AdminPrefixUrl + areaName + folderPath `

Here I am trying to replace `areaName + folderPath ` with  `folderRoute ` but not used `String.Replace` to take care of `/`, If its Admin route then, `areaName + folderPath ` will be somewhere in the middle not at start of the template.

ns8482e commented on src/OrchardCore.Modules/OrchardCore.Admin/AdminFilter.cs:
```diff
@@ -24,25 +24,44 @@ public override async Task OnActionExecutionAsync(ActionExecutingContext context
         }
 
-        private bool IsNameAdmin(ActionExecutingContext context)
+        private async Task<bool?> IsAdminAuthorizedAsync(Microsoft.AspNetCore.Http.HttpContext context)
```
I used `bool?` for a purpose, It returns null if `AdminAttribute` is not applied, can continue calling next () 
true -  AdminAttribute is applied and user authorized to access admin panel and continue calling next()
false - AdminAttribute is applied and user is not authorized and can't access admin panel

ns8482e commented on src/OrchardCore.Modules/OrchardCore.Admin/AdminFilter.cs:
```diff
@@ -24,25 +24,44 @@ public override async Task OnActionExecutionAsync(ActionExecutingContext context
-            return string.Equals(context.Controller.GetType().Name, "Admin", StringComparison.OrdinalIgnoreCase);
+            if (AdminAttribute.IsApplied(context))
+                return await _authorizationService.AuthorizeAsync(context.User, Permissions.AccessAdminPanel);
+            return null;            
```
null to identify if AdminAttribute is not applied, so I'll change it to
`return Task.FromResult(null)`

ns8482e commented on src/OrchardCore/OrchardCore.Mvc.Core/RazorPages/PageConventionCollectionExtensions.cs:
```diff
@@ -17,13 +17,14 @@ public static class PageConventionCollectionExtensions
                 {
-                    if (selector.AttributeRouteModel.Template.StartsWith(areaName, StringComparison.Ordinal))
+                    var path = $"{areaName}/{folderPath.Trim('/')}";
+                    if (selector.AttributeRouteModel.Template.Contains( path, StringComparison.Ordinal))
```
> Hmm, i would have kept my code, the conventions are that area never have a slash, folderPath should start with a slash and folderRoute shouldn't start with a slash

Your code worked until I added, 
`public class AdminFolderModelConvention : IPageRouteModelConvention`

Template no longer starting with `areaName + folderPath ` now it starts with `AdminPrefixUrl + areaName + folderPath `

ns8482e commented on src/OrchardCore.Modules/OrchardCore.Admin/AdminFilter.cs:
```diff
@@ -24,25 +24,44 @@ public override async Task OnActionExecutionAsync(ActionExecutingContext context
-            }
+            }                        
+
+            if ( (await IsAdminAuthorizedAsync(context.HttpContext)) == false )
```
I am using Visual Studio Code. Thanks for the Tip.

I used bool? to differentiate between true and `AdminAttribute` not applied, If it seems not useful then I'll change it to `bool`

ns8482e commented on src/OrchardCore.Modules/OrchardCore.Admin/AdminFolderModelConvention.cs:
```diff
@@ -0,0 +1,28 @@
+                for (var i = 0; i <  model.Selectors.Count; i++ )
+                {
+                    var selector = model.Selectors[i];                    
+                    var template = selector.AttributeRouteModel.Template.Replace("/Admin/","/", StringComparison.OrdinalIgnoreCase);                    
```
This is only for Razor Pages, If Template contains `/Admin/` then it means template is default template based on Admin folder. e.g.

```
/Pages/Admin/Create.cshtml => [area]/Admin/Create
/Pages/Foo/Admin/Edit.cshtml => [area]/Foo/Admin/Edit
```

Here above default template will be replaced as following by removing `/Admin/` folder from path.

```
/Admin/[area]/Create
/Admin/[area]/Foo/Edit
``` 
`asp-page="/Admin/Create"`  will create link as `/Admin/[area]/Create`

jtkech commented:
Okay i understand, need to try it to fully understand what happens.

I will fork your branch, then at least i will help you to format the files and commit directly on your branch. Then maybe i will do some other changes that you will be able to review ;)

jtkech commented:
I don't have permissions to push to your repo so i will upload here some zip files with some updates and reformatting.

- First update in [AdminFilter.zip](https://github.com/OrchardCMS/OrchardCore/files/4069054/AdminFilter.zip) where i tried and just used the following returning a `Task<bool>` directly without async await keywords.

        if (!await AuthorizeAsync(context.HttpContext))

        ...

        private Task<bool> AuthorizeAsync(HttpContext context)
        {
            if (AdminAttribute.IsApplied(context))
            {
                return _authorizationService.AuthorizeAsync(context.User, ...);
            }

            return Task.FromResult(true);
        }

- Update and formatting of [AdminZoneFilter.zip](https://github.com/OrchardCMS/OrchardCore/files/4069221/AdminZoneFilter.zip).

- Update [AdminFolderModelConvention.zip](https://github.com/OrchardCMS/OrchardCore/files/4069507/AdminFolderModelConvention.zip), good job, here we have only one template per selector and for an admin page that's okay to replace this default template with one starting with the prefix. I didn't know about `CombineTemplates()` that provides a `CleanTemplate` ;) 

- More coming later or tomorrow ;) For infos i updated the list to only show latest items (here we never publish them) so that when you delete an item it is no more shown in the list.

    Need to think about `AddAreaFolderRoute()` for admin page routes (by folder), we need to still allow people to redefine a folder / route without the admin prefix. I started to test an additional `AddAdminAreaFolderRoute()` on which we provide the admin prefix, will see tomorrow.

ns8482e commented on src/OrchardCore.Modules/OrchardCore.Demo/Pages/Foo/Admin/Edit.cshtml:
```diff
@@ -0,0 +1,110 @@
+@using OrchardCore.ContentManagement.Display;
+@using OrchardCore.DisplayManagement.ModelBinding;
+@using OrchardCore.ContentManagement.Records;
+@using System.ComponentModel.DataAnnotations;
```
Ok -i'll change the order, However I observed in OC that most using are in alphabetic order

ns8482e commented:
@jtkech  Changes made per your review comments.

> * For infos i updated the list to only show latest items (here we never publish them) so that when you delete an item it is no more shown in the list.

Fixed, by adding Published condition

> * I didn't know about `CombineTemplates()` that provides a `CleanTemplate` ;)

 [Thanks to open source](https://github.com/dotnet/aspnetcore/blob/f79f2e3b1200f8e672b77583a54e6157e49da9e4/src/Mvc/Mvc.Core/src/ApplicationModels/AttributeRouteModel.cs#L145)

ns8482e commented on src/OrchardCore.Modules/OrchardCore.Admin/AdminZoneFilter.cs:
```diff
@@ -1,29 +1,34 @@
             }
+            else if(context.ActionDescriptor is PageActionDescriptor page)
+            {
+                if( page.ViewEnginePath.Contains("/Admin/", StringComparison.OrdinalIgnoreCase))
```
Using `Contains`

jtkech commented:
@ns8482e thanks for your updates

- About usings ordering, yes it is alphabetic unless system ones that are put at the top, that's the convention used by Microsoft. While editing under VStudio, if you right click on the usings and then do a `Remove and Sort Usings` it removes unecessary usings, sort them by alphabetic order but put the system ones at the top.

- I will fetch your changes but i have some others on my side, it would have been good if i had the permissions to push them in your branch directly.

- As i remember i fixed a folder route in the demo startup, init somes var with a shared value before a foreach loop, i will remove the usage of `AdminActionConstraint` that has been removed in #5255.

- But the last main point is about `AddAreaFolderRoute()` for admin routes, for testing i did an `AddAdminAreaFolderRoute` with an prefix parameter. But the problem is that our new convention that prefixes admin routes runs too soon, it would have to be added last. As we do in #5255 for controller in `OC.Mvc startup Configure()` that runs last because of its `ConfigureOrder` (but its `ConfigureServices()` runs first). But `OC.Mvc` is not aware of the admin prefix, that's why it resolves services whose one implementation is defined in `OC.Admin`.

- So, i will finalize this this night.

sebastienros commented:
@ns8482e I can edit your PR, have you removed this restriction recently or is it @jtkech who must have made a mimstake when he tried?

jtkech commented:
Just retried a testing change.

    "C:\Program Files\Git\bin\git.exe" push --recurse-submodules=check --progress "origin" 
    refs/heads/ns8482e/#3739:refs/heads/ns8482e/#3739
    remote: Permission to surevelox/OrchardCore.git denied to jtkech.
    fatal: unable to access 'https://github.com/surevelox/OrchardCore.git/': The requested URL 
    returned error: 403
    Done

    Press Enter or Esc to close console...

sebastienros commented:
I usually do changes directly from the github interface

jtkech commented:
Ah okay, i will try to do it this way after i tried my changes locally.

Yes because i already have 7 updated files just for formatting, so would not be easy to share and i don't want to bother @ns8482e with these formatting rules ;)

![Formatting](https://user-images.githubusercontent.com/8586360/72572619-63101d00-38c3-11ea-80b2-b535b207a762.png)

ns8482e commented:
@sebastienros @jtkech in past when I create PR I always have _allow edits from maintainers_ checked.  It seems something has changed in github.

jtkech commented:
@ns8482e no worries ;)

- So, i worked on it, so in my local branch now the `AdminFolderModelConvention` that generates prefixed templates is executed last by registering it in `OC.Admin` but in a specific startup with an high `Order`. So the code in `AddAreaFolderRoute()` can work as before. `AddAreaFolderRoute()` is now only applied to non admin page paths, and i did a new `AddAdminAreaFolderRoute()` that is only applied to admin page paths containing `/Admin/`.

- So now, for a page path that contains `/Admin/` we don't have any default route that is not prefixed with the admin prefix, this unless we explicitly register one without the prefix and that takes precedence for link generation. All seems to work fine.

- But but but, i still have a last concern when the path doesn't contain `/Admin/` but the `AdminAttribute` is used on the `PageModel`, i would like to also prefix the default route with the admin prefix, as done in `AdminFolderModelConvention`, but at this stage there is no page model on which we could check if an admin attribute is applied.

- Who decide to prefix admin default routes ;)

So, i need to work on it a little more, so wait before doing more changes. Right now i will not push my changes as zip files as it is a pain. So yes would be great if you could check the permissions on your repo so that i could push my changes tomorrow, thanks ;)

ns8482e commented:
@jtkech @sebastienros  if you can edit using github interface, then you can push changes using github desktop.

jtkech commented:
So, to see if a page model is decorated with the `[Admin]` attribute, we would need the page type. In dev mode we don't use precompiled views so at the time we apply a convention the pages are not yet compiled. We could use the `PageLoader` but that will need to compile all pages in one time.

We could retrieve the view descriptors from the `.Views.dll`, as we do to adapt their relative paths when they are used (only in prod mode). But here we are not aware of the admin prefix, we would need an abstraction, we read them only in prod mode, and it would be still quite complex.

So seems too complex to check **in a convention** if a page model has an admin attribute, so for such a page, that is already protected, if we still want an auto-prefixed default route, the convention would be that the page file name starts with `Admin`.

As described in the following comments

    // An admin page is a page whose path contains an '/Admin/' segment or whose model has been decorated
    // with the [Admin] attribute. An admin page access is protected, the user needs to be authenticated.

    // A path is recognized as an admin path if it contains an '/Admin/' segment or if the file name starts
    // with 'Admin'. The default path based route of a page having an admin path is automatically prefixed.

    // So an admin page using the [Admin] attribute may not have an admin path, if you still want a default
    // route that is automatically prefixed, the convention is that the page name has to start with 'Admin'.

jtkech commented:
So i can edit your repo but uploads are disabled.

    File uploads require push access to this repository.

But that's okay i've not so many files to update ;)

ns8482e commented on src/OrchardCore.Modules/OrchardCore.Admin/Startup.cs:
```diff
@@ -56,4 +57,24 @@ public override void Configure(IApplicationBuilder builder, IEndpointRouteBuilde
         }
     }
+
+    public class AdminPagesStartup : StartupBase
```
Separate startup? without feature attribute?

jtkech replied:
I will comment it

ns8482e commented:
I think, for Admin replacing the original route makes more sense then just adding new one as Front end.

Also for usability perspective, I believe if someone is configuring `AdminUrlPrefix `that means he want to append `/Admin/` for all their admin routes - regardless of module Startup configuration should not require each route be configured individually for AdminUrlPrefix ,  after all, Web Application is super module.

e.g.  `/Page/Foo/Admin/Edit.cshtml`  MVC default route is `/Foo/Admin/Edit`, configuring `AdminUrlPrefix = xyz`,  should automatically adjust default route to `/xxx/Foo/Edit`

Requiring to configure every route with  `AdminUrlPrefix ` is useless, if that's the case then user can define, his own configuration value and why use - `AdminOptions.AdminUrlPrefix ` doesn't add any value  except a placeholder then.

ns8482e commented:
> when you define a route through the `@page directive` it will replace this default route, and we should not update it because it has been explicitly set by the `@page` directive

It depends,  if `AdminPrefixUrl = "xyz"`

If Page is `/Pages/Foo/Admin/Edit.cshtml`  the default MVC route is `/xyz/Foo/Edit` with `AdminPageModelConversion`

If user configures `@page "myfoo/{id}"`, it should be`/xyz/{area}/Foo/Edit/myfoo/{id}`
if user configures `@page "~/myfoo/{id}"` it should be`/xyz/myfoo/{id}`

My understanding is defining `~/` is page url is from `root` here `xyz` is root for admin :)

jtkech commented:
@ns8482e 

> Should generate `/Manage/Foo/Edit/4b2cwr0evtermrmxn02bejc5nh`

This is what it does but in the demo startup i added this example

    options.Conventions.AddAdminAreaFolderRoute("OrchardCore.Demo", "/", _adminOptions.AdminUrlPrefix);

That applies a global folder convention for all admin paths under the root of the module `/` but that overrides the convention on `/Foo/Admin`. If we remove this line it works as expected, i will just remove this line in the sample code.

> Separate startup? without feature attribute?

Yes we can, it is then tied to the module main feature as the original startup. This allows to apply different orders (`Order` and / or `ConfigureOrder`) without affecting the ordering of the original startup that naturally depends on the features dependencies.

> I think, for Admin replacing the original route makes more sense then just adding new one as Front end.

> Also for usability perspective, I believe if someone is configuring `AdminUrlPrefix `that means he want to append `/Admin/` for all their admin routes - regardless of module Startup configuration. He need not require each route be configured individually using AdminUrlPrefix ,  after all, Web Application is super module.

> e.g.  `/Page/Foo/Admin/Edit.cshtml`  MVC default route is `/Foo/Admin/Edit`, configuring `AdminUrlPrefix = xyz`,  should automatically adjust default route to `/xxx/Foo/Edit`

> Requiring to configure every route with  `AdminUrlPrefix ` is useless, if that's the case then user can define, his own configuration value and why would use - `AdminOptions.AdminUrlPrefix `? It doesn't add any value  except being a placeholder variable.

We already adjust the MVC default route to `{module = area}/Foo/Admin/Edit`. I understand your suggestion to atomatically prefix all admin paths but, without saying what is the best, i'm following the rules that has been discussed and applied in #5225 saying that the default route is automatically prefixed, so here `/xxx/{area}/Foo/Edit`, but not those that are explicitly defined.

> when you define a route through the `@page directive` it will replace this default route, and we should not update it because it has been explicitly set by the `@page` directive

> > It depends,  if `AdminPrefixUrl = "xyz"`

> > If Page is `/Pages/Foo/Admin/Edit.cshtml`  the default MVC route is `/xyz/Foo/Edit` with `AdminPageModelConversion`

> > If user configures `@page "myfoo/{id}"`, it should be`/xyz/{area}/Foo/Edit/myfoo/{id}`
if user configures `@page "~/myfoo/{id}"` it should be`/xyz/myfoo/{id}`

> > My understanding is defining `~/` is page url is from `root` here `xyz` is root for admin :)

Same answer, here `/xyz/{area}/Foo/Edit/myfoo/{id}` would be the default route automatically prefixed if the `@page` doesn't define one and that **replaces** the default one. For `@page "/myfoo/{id}"` and `@page "~/myfoo/{id}"` mvc generates the same `myfoo/{id}` template. Then for the relative `@page "myfoo/{id}"` it generates `{module = area}/{page base path}/myfoo/{id}`.

 Again i understand your suggestion, but the idea is to only prefix the default route (unless overriden by the `@page` directive) and let people defining additional routes as they want with or without a prefix.

Finally, as a last try, i'm still trying to check in the convention if the page model is decorated with the admin attribute, at least in production where pages are precompiled.

jtkech commented:
@ns8482e 

- I had a discussion about the admin prefix, as it is done in #5225 to avoid path conflicts we don't remove the `/Admin/` segment of an admin path, we just prefix it with the current admin prefix. So it is simpler as we don't need anymore to remove this segment.

- I found a way to retrieve the admin attribute from the page view descriptors so that, to automatically prefix the default route, we don't need anymore to check if the page name starts with `Admin`. The only restriction is that if you add / remove the admin attribute live, you need to restart the app. Otherwise, as before you can do other changes live in the page.

I will commit my changes tomorrow

Sorry for the long story, you did a good job and most of the work about protecting admin pages detected by path or by the admin attribute, this is just the convention related to the usage of the admin prefix that was not so clear, and then the way to retrieve the admin attribute in a `IPageRouteModelConvention` that was not so simple. but this PR will be completed soon ;)

jtkech commented:
@ns8482e

I'm done with my changes, so i kept most of your original changes to make the admin razor pages protected, and i mainly adapted the way we automatically prefix the default route of an admin page.

The last part could have been done in a separate PR or in the pending PR #5225 that does the same thing but for controllers, but i thought that it was more related to this PR.

I think that both #5225 and this PR will have to be merged together.

jtkech commented:
Okay cool, just merged
